crates/iceberg/src/arrow.rs

```diff
@@ -106,3 +114,732 @@ impl ArrowReader {
+            field_type: Box::new(value.clone()),
+            initial_default: None,
+            write_default: None,
+        });
+        Ok(Type::List(ListType { element_field }))
+    }
+
+    fn map(&mut self, map: &DataType, key_value: Self::T, value: Self::T) -> Result<Self::T> {
+        match map {
+            DataType::Map(field, _) => match field.data_type() {
+                DataType::Struct(fields) => {
+                    if fields.len() != 2 {
+                        return Err(Error::new(
+                            ErrorKind::DataInvalid,
+                            "Map field must have exactly 2 fields",
+                        ));
+                    }
+
+                    let key_field = &fields[0];
+                    let value_field = &fields[1];
+
+                    let key_id = get_field_id(key_field)?;
+                    let key_doc = get_field_doc(key_field);
+                    let key_field = Arc::new(NestedField {
+                        id: key_id,
+                        doc: key_doc,
+                        name: "key".to_string(),
+                        required: !key_field.is_nullable(),
+                        field_type: Box::new(key_value.clone()),
+                        initial_default: None,
+                        write_default: None,
+                    });
+
+                    let value_id = get_field_id(value_field)?;
+                    let value_doc = get_field_doc(value_field);
+                    let value_field = Arc::new(NestedField {
+                        id: value_id,
+                        doc: value_doc,
+                        name: "value".to_string(),
+                        required: !value_field.is_nullable(),
+                        field_type: Box::new(value.clone()),
+                        initial_default: None,
+                        write_default: None,
+                    });
+
+                    Ok(Type::Map(MapType {
+                        key_field,
+                        value_field,
+                    }))
+                }
+                _ => Err(Error::new(
+                    ErrorKind::DataInvalid,
+                    "Map field must have struct type",
+                )),
+            },
+            _ => Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Map type must have map data type",
+            )),
+        }
+    }
+
+    fn primitive(&mut self, p: &DataType) -> Result<Self::T> {
+        match p {
+            DataType::Boolean => Ok(Type::Primitive(PrimitiveType::Boolean)),
+            DataType::Int32 => Ok(Type::Primitive(PrimitiveType::Int)),
+            DataType::Int64 => Ok(Type::Primitive(PrimitiveType::Long)),
+            DataType::Float32 => Ok(Type::Primitive(PrimitiveType::Float)),
+            DataType::Float64 => Ok(Type::Primitive(PrimitiveType::Double)),
+            DataType::Decimal128(p, s) => Ok(Type::Primitive(PrimitiveType::Decimal {
```

Comment:
Use `Type::decimal` instead? iceberg has some limitations for decimal's precision and scale.